File: src/cfdx/io/cfdx_io/gap_analysis.cpp
// M0.10-T03: GapAnalysis report generation (Markdown + JSON)
//
// Spécification CFDX v0.7 issue #425 §Gap Analysis Engine
// ---------------------------------------------------------------------------
// Generates human-readable Markdown and machine-readable JSON reports
// from the accumulated findings of a conversion.
#include "gap_analysis.h"
#include "io_interface.h"

#include <sstream>
#include <iomanip>

namespace cfdx {
namespace io {

// ---------------------------------------------------------------------------
// JSON serialisation (manual, no external dependency)
// ---------------------------------------------------------------------------
static std::string json_escape(const std::string& s) {
    std::string out;
    out.reserve(s.size() + 8);
    for (char c : s) {
        switch (c) {
            case '"':  out += "\\\""; break;
            case '\\': out += "\\\\"; break;
            case '\n': out += "\\n";  break;
            case '\r': out += "\\r";  break;
            case '\t': out += "\\t";  break;
            default:   out += c;      break;
        }
    }
    return out;
}

std::string GapAnalysis::to_json() const {
    std::ostringstream ss;
    ss << std::boolalpha;
    ss << "{\n";
    ss << "  \"has_blocking\": " << has_blocking() << ",\n";
    ss << "  \"summary\": {\n";
    ss << "    \"supported\": " << n_supported() << ",\n";
    ss << "    \"approximated\": " << n_approximated() << ",\n";
    ss << "    \"unsupported_nonblocking\": " << n_unsupported_nonblocking() << ",\n";
    ss << "    \"unsupported_blocking\": " << n_unsupported_blocking() << ",\n";
    ss << "    \"unavailable\": " << n_unavailable() << ",\n";
    ss << "    \"total\": " << findings_.size() << "\n";
    ss << "  },\n";
    ss << "  \"findings\": [\n";
    for (std::size_t i = 0; i < findings_.size(); ++i) {
        const auto& f = findings_[i];
        ss << "    {\n";
        ss << "      \"severity\": \"" << to_string(f.severity) << "\",\n";
        ss << "      \"category\": \"" << json_escape(f.category) << "\",\n";
        ss << "      \"feature\": \"" << json_escape(f.feature) << "\",\n";
        ss << "      \"detail\": \"" << json_escape(f.detail) << "\",\n";
        ss << "      \"suggestion\": \"" << json_escape(f.suggestion) << "\"\n";
        ss << "    }";
        if (i + 1 < findings_.size()) ss << ",";
        ss << "\n";
    }
    ss << "  ]\n";
    ss << "}\n";
    return ss.str();
}

// ---------------------------------------------------------------------------
// Markdown serialisation
// ---------------------------------------------------------------------------
std::string GapAnalysis::to_markdown() const {
    std::ostringstream ss;

    ss << "# Gap Analysis Report\n\n";
    ss << "## Summary\n\n";
    ss << "| Severity | Count |\n";
    ss << "|----------|-------|\n";
    ss << "| Supported / Mapped | " << n_supported() << " |\n";
    ss << "| Approximated | " << n_approximated() << " |\n";
    ss << "| Unsupported (non-blocking) | " << n_unsupported_nonblocking() << " |\n";
    ss << "| Unsupported (blocking) | " << n_unsupported_blocking() << " |\n";
    ss << "| Unavailable (source info missing) | " << n_unavailable() << " |\n";
    ss << "| **Total** | " << findings_.size() << " |\n";
    ss << "\n";
    ss << "**Blocking incompatibilities:** "
       << (has_blocking() ? "YES — conversion blocked" : "NONE") << "\n\n";

    // Group by severity for readability
    static const struct { Severity sev; const char* header; } categories[] = {
        {Severity::SUPPORTED,             "## 1. Supported / Mapped Features\n"},
        {Severity::APPROXIMATED,          "## 2. Approximated / Fallback Mappings\n"},
        {Severity::UNAVAILABLE,           "## 3. Unavailable Source Information\n"},
        {Severity::UNSUPPORTED_NONBLOCK,  "## 4. Unsupported Features (non-blocking)\n"},
        {Severity::UNSUPPORTED_BLOCK,     "## 5. Unsupported Features (blocking)\n"},
    };

    for (const auto& cat : categories) {
        bool any = false;
        for (const auto& f : findings_) {
            if (f.severity == cat.sev) {
                if (!any) {
                    ss << cat.header;
                    ss << "\n| Category | Feature | Detail | Suggestion |\n";
                    ss << "|----------|---------|--------|------------|\n";
                    any = true;
                }
                ss << "| " << f.category << " | " << f.feature
                   << " | " << f.detail << " | " << (f.suggestion.empty() ? "—" : f.suggestion) << " |\n";
            }
        }
        if (!any) {
            ss << cat.header;
            ss << "\n_None._\n";
        }
        ss << "\n";
    }

    return ss.str();
}

}  // namespace io
}  // namespace cfdx

File: src/cfdx/python/cfdx/io/gap_analysis.py
"""Gap Analysis report engine — generates Markdown + JSON reports.

Corresponds to the C++ header:
  src/cfdx_io/gap_analysis.h/.cpp
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Optional

from cfdx.io.schema import Severity


@dataclass
class Finding:
    severity: Severity
    category: str
    feature: str
    detail: str = ""
    suggestion: str = ""


class GapAnalysis:
    """Accumulates findings from a conversion/import operation.

    Every unsupported or approximated feature must be recorded here.
    No silent substitutions/fallbacks.
    """

    def __init__(self) -> None:
        self._findings: list[Finding] = []

    def add(
        self,
        severity: Severity,
        category: str,
        feature: str,
        detail: str = "",
        suggestion: str = "",
    ) -> "GapAnalysis":
        self._findings.append(
            Finding(severity, category, feature, detail, suggestion)
        )
        return self

    def supported(self, category: str, feature: str, detail: str = "") -> "GapAnalysis":
        return self.add(Severity.SUPPORTED, category, feature, detail)

    def approximated(self, category: str, feature: str, detail: str, suggestion: str = "") -> "GapAnalysis":
        return self.add(Severity.APPROXIMATED, category, feature, detail, suggestion)

    def unsupported_nonblocking(self, category: str, feature: str, detail: str, suggestion: str = "") -> "GapAnalysis":
        return self.add(Severity.UNSUPPORTED_NONBLOCK, category, feature, detail, suggestion)

    def unsupported_blocking(self, category: str, feature: str, detail: str, suggestion: str = "") -> "GapAnalysis":
        return self.add(Severity.UNSUPPORTED_BLOCK, category, feature, detail, suggestion)

    def unavailable(self, category: str, feature: str, detail: str) -> "GapAnalysis":
        return self.add(Severity.UNAVAILABLE, category, feature, detail)

    @property
    def findings(self) -> list[Finding]:
        return list(self._findings)

    def has_blocking(self) -> bool:
        return any(f.severity == Severity.UNSUPPORTED_BLOCK for f in self._findings)

    def n_supported(self) -> int:
        return sum(1 for f in self._findings if f.severity in (Severity.SUPPORTED, Severity.APPROXIMATED))

    def n_approximated(self) -> int:
        return sum(1 for f in self._findings if f.severity == Severity.APPROXIMATED)

    def n_unsupported_nonblocking(self) -> int:
        return sum(1 for f in self._findings if f.severity == Severity.UNSUPPORTED_NONBLOCK)

    def n_unsupported_blocking(self) -> int:
        return sum(1 for f in self._findings if f.severity == Severity.UNSUPPORTED_BLOCK)

    def n_unavailable(self) -> int:
        return sum(1 for f in self._findings if f.severity == Severity.UNAVAILABLE)


class GapAnalysisReport:
    """Generates Markdown and JSON reports from a GapAnalysis."""

    def __init__(self, gap: GapAnalysis) -> None:
        self.gap = gap

    def to_json(self) -> str:
        data = {
            "has_blocking": self.gap.has_blocking(),
            "summary": {
                "supported": self.gap.n_supported(),
                "approximated": self.gap.n_approximated(),
                "unsupported_nonblocking": self.gap.n_unsupported_nonblocking(),
                "unsupported_blocking": self.gap.n_unsupported_blocking(),
                "unavailable": self.gap.n_unavailable(),
                "total": len(self.gap.findings),
            },
            "findings": [
                {
                    "severity": f.severity.value,
                    "category": f.category,
                    "feature": f.feature,
                    "detail": f.detail,
                    "suggestion": f.suggestion,
                }
                for f in self.gap.findings
            ],
        }
        return json.dumps(data, indent=2)

    def to_markdown(self) -> str:
        lines = ["# Gap Analysis Report", ""]
        lines.append("## Summary")
        lines.append("")
        lines.append("| Severity | Count |")
        lines.append("|----------|-------|")
        lines.append(f"| Supported / Mapped | {self.gap.n_supported()} |")
        lines.append(f"| Approximated | {self.gap.n_approximated()} |")
        lines.append(f"| Unsupported (non-blocking) | {self.gap.n_unsupported_nonblocking()} |")
        lines.append(f"| Unsupported (blocking) | {self.gap.n_unsupported_blocking()} |")
        lines.append(f"| Unavailable (source info missing) | {self.gap.n_unavailable()} |")
        lines.append(f"| **Total** | {len(self.gap.findings)} |")
        lines.append("")

        if self.gap.has_blocking():
            lines.append("**Blocking incompatibilities:** YES — conversion blocked")
        else:
            lines.append("**Blocking incompatibilities:** NONE")
        lines.append("")

        categories = [
            ("## 1. Supported / Mapped Features", Severity.SUPPORTED),
            ("## 2. Approximated / Fallback Mappings", Severity.APPROXIMATED),
            ("## 3. Unavailable Source Information", Severity.UNAVAILABLE),
            ("## 4. Unsupported Features (non-blocking)", Severity.UNSUPPORTED_NONBLOCK),
            ("## 5. Unsupported Features (blocking)", Severity.UNSUPPORTED_BLOCK),
        ]

        for header, sev in categories:
            items = [f for f in self.gap.findings if f.severity == sev]
            lines.append(header)
            lines.append("")
            if items:
                lines.append("| Category | Feature | Detail | Suggestion |")
                lines.append("|----------|---------|--------|------------|")
                for f in items:
                    lines.append(
                        f"| {f.category} | {f.feature} | {f.detail} | "
                        f"{f.suggestion if f.suggestion else '—'} |"
                    )
            else:
                lines.append("_None._")
            lines.append("")

        return "\n".join(lines)
